### DISCUSSION TITLE
Inference on many videos in Google Colab

### DISCUSSION THREAD

**Discussion by amblypatty** [talmolab/sleap#1555] [Help!] [answered]
Hello,

I am having an issue trying to get Google Colab to run top-down inference on many videos within a folder in my Google Drive that is mounted. Below is the bash script I have made to iterate through the videos in the video folder and perform inference and tracking on each video.
```
%%writefile sleap_processing.sh
#!/bin/bash

# Define the path to the SLEAP executable and model paths
sleap_executable="sleap-track"
model_paths="-m 'models/date.centroid' -m 'models/date.centered_instance'"

# Define tracking parameters
tracking_params="--tracking.tracker flowmaxtracks --tracking.max_tracking 1 --tracking.max_tracks 1 --tracking.post_connect_single_breaks 1 --tracking.similarity instance --tracking.window 5 --tracking.match greedy"

# Get the list of video files in Google Drive folder
video_folder="/content/drive/MyDrive/video_folder"
video_files=$(ls "$video_folder")

# Iterate through the video files and run SLEAP commands
for video_file in $video_files; do
    if [[ $video_file == *.mp4 ]]; then
        input_video_path="$video_folder/$video_file"
        output_sleap_path="/content/drive/MyDrive/output_predictions_$video_file.slp"
        command="$sleap_executable $model_paths $tracking_params -o $output_sleap_path $input_video_path"
        echo "Processing $input_video_path..."
        $command
    fi
done
```
Yet, I get a `ValueError` when I run this command stating that it could not detect the backend for the filename of 5, which I think tells me that the command is using an index number of a video as the video name. Here is the error output:
```
Versions:
SLEAP: 1.3.3
TensorFlow: 2.8.4
Numpy: 1.22.4
Python: 3.10.12
OS: Linux-5.15.120+-x86_64-with-glibc2.35

System:
GPUs: 1/1 available
  Device: /physical_device:GPU:0
         Available: True
        Initalized: False
     Memory growth: True

Video: 5
Traceback (most recent call last):
  File "/usr/local/bin/sleap-track", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.10/dist-packages/sleap/nn/inference.py", line 5409, in main
    provider, data_path = _make_provider_from_cli(args)
  File "/usr/local/lib/python3.10/dist-packages/sleap/nn/inference.py", line 5271, in _make_provider_from_cli
    provider = VideoReader.from_filepath(
  File "/usr/local/lib/python3.10/dist-packages/sleap/nn/data/providers.py", line 344, in from_filepath
    video = sleap.Video.from_filename(filename, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/sleap/io/video.py", line 1276, in from_filename
    raise ValueError(
ValueError: Could not detect backend for specified filename: 5
```

However, when I view the `video_files`, the list of video names in the folder all populate appropriately, so I am unsure of how to get SLEAP to access the string name in the list. Additionally, the first iteration is using `video: 5`, but the folder has only .mp4 video files in it, so that confuses me more. 
```
video_folder="/content/drive/MyDrive/video_folder"
video_files=!(ls "$video_folder")
video_files
```
```
['video1.mp4',
 'video2.mp4',
 'video3.mp4',
...
 'video60.mp4',
 'video61.mp4',
 'video62.mp4']
```
Can you help me fix this so that I don't have to run a single inference script on over 60 videos manually?

**Comment by roomrys** [comment, marked as answer]
Hi @amblypatty,

Do you mind echoing the command that your script is calling before executing so we can see what's been stitched together?

Thanks,
Liezl

**Reply by amblypatty** [reply]
Thank you for asking that. Once I figured out how to echo the command with running: `!bash -x sleap_processing.sh`, I found that I switched the `$model_paths` and the `$input_video_path` which made `sleap-track` use nothing as the input. Additionally, the `$model_paths` did not need to be individually quoted inside the main quotations.
